app/Utils/ClientPortal/CustomMessage/CustomMessage.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+namespace App\Utils\ClientPortal\CustomMessage;
+
+use App\Models\Client;
+use App\Models\ClientContact;
+use App\Models\Company;
+use App\Models\GroupSetting;
+
+class CustomMessage
+{
+    protected $message;
+
+    protected $values;
+
+    protected $client;
+
+    protected $company;
+
+    protected $contact;
+
+    protected $group;
+
+    protected $entity;
+
+    protected $invitation;
+
+    public function client(Client $client): self
+    {
+        $this->client = $client;
+
+        return $this;
+    }
+
+    public function company(Company $company): self
+    {
+        $this->company = $company;
+
+        return $this;
+    }
+
+    public function contact(ClientContact $contact): self
+    {
+        $this->contact = $contact;
+
+        return $this;
+    }
+
+    public function group(GroupSetting $group): self
+    {
+        $this->group = $group;
+
+        return $this;
+    }
+
+    public function entity($entity): self
+    {
+        $this->entity = $entity;
+
+        return $this;
+    }
+
+    public function invitation($invitation): self
+    {
+        $this->invitation = $invitation;
+
+        return $this;
+    }
+
+    public function message(string $message): string
+    {
+        $this->message = $message;
+        $this->values = $this->compose();
+
+        return strtr($message, $this->values);
+    }
+
+    private function compose(): array
+    {
+        return [
+            '$company.id' => optional($this->company)->id,
+            '$company.name' => optional($this->company)->getSetting('name'),
+            '$company.website' => optional($this->company)->getSetting('website'),
+
+            '$client.id' => optional(optional($this->client)->present())->id,
+            '$client.name' => optional(optional($this->client)->present())->name,
+            '$client.website' => optional(optional($this->client)->present())->website,
+            '$client.public_notes' => optional(optional($this->client)->present())->public_notes,
+            '$client.phone' => optional(optional($this->client)->present())->phone,
```

Comment:
If you are just showing the variables, no need to ->present() them, just use the variable name ie. this->client->website

The presenter is mainly used for fields like ->address() which bundles the street/city/postal code etc etc.

Reply:
Fixed here: https://github.com/invoiceninja/invoiceninja/pull/3626